### PR TITLE
[Preprocess #1] Business logic for FinnGen summary stats ingestion

### PR DESCRIPTION
**Pull request by tskir**
Closes https://github.com/opentargets/issues/issues/3095. This is the first part of breaking down https://github.com/opentargets/genetics_etl_python/pull/144 into smaller PRs for review.

### CODE REVIEW

**Comment by tskir** [comment]
Hi @d0choa @DSuveges, I've rebased against recent changes in main. Tests for FinnGen sumstat ingestion have been added and are passing. Ready for review :zap:

File: src/otg/preprocess/finngen/study_index.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
+"""FinnGen ingestion: study index."""
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/otg/preprocess/finngen/study_index.py:1]: More specifically, I'm not sure if I'm getting the difference between preprocess and datasource.

File: src/otg/preprocess/finngen/study_index.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                    "studyId"
+                ),
+                f.col("phenostring").alias("traitFromSource"),
+                f.col("num_cases").alias("nCases"),
+                f.col("num_controls").alias("nControls"),
+                (f.col("num_cases") + f.col("num_controls")).alias("nSamples"),
+                f.lit(finngen_release_prefix).alias("projectId"),
+                f.lit("gwas").alias("studyType"),
+                f.lit(True).alias("hasSumstats"),
+                f.lit("377,277 (210,870 females and 166,407 males)").alias(
```
Comment (DSuveges) [inline src/otg/preprocess/finngen/study_index.py:74]: Some of the recently added logic from main is missing here:
- Adding column ["discoverySamples"](https://github.com/opentargets/genetics_etl_python/blob/f60e7205ea1c4306b85099e9fcb4e149b3ec5e1d/src/otg/datasource/finngen/study_index.py#L60)
- Adding column ["ldPopulationStructure"](https://github.com/opentargets/genetics_etl_python/blob/f60e7205ea1c4306b85099e9fcb4e149b3ec5e1d/src/otg/datasource/finngen/study_index.py#L76)

File: src/otg/preprocess/finngen/study_index.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    - Study ID in the special format (FINNGEN_R9_*)
+    - Trait name (for example, Amoebiasis)
+    - Number of cases and controls
+    - Link to the summary statistics location
+
+    Some fields are also populated as constants, such as study type and the initial sample size.
+    """
+
+    @classmethod
+    def get_schema(cls: type[FinnGenStudyIndex]) -> StructType:
```
Comment (d0choa) [inline src/otg/preprocess/finngen/study_index.py:39]: You don't need this function. You get it by inheritance from StudyIndex

File: src/otg/preprocess/finngen/study_index.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                    f.lit(finngen_summary_stats_url_prefix),
+                    f.col("phenocode"),
+                    f.lit(finngen_summary_stats_url_suffix),
+                ).alias("summarystatsLocation"),
+            ),
+            _schema=cls.get_schema(),
+        )
+
+
+def ingest_finngen_study_index(
```
Comment (d0choa) [inline src/otg/preprocess/finngen/study_index.py:87]: we are placing this logic in (e.g) `src/otg/finngen.py`. They are the definition of the "steps" and usually contain the minimum number of function calls to get the ETL functionality: read datasets, transform them, write the outputs. This separation of concerns helps to decouple the data model from the actual execution. For example, we pretty much don't have a spark session in any `otg/dataset`, `otg/method` or `otg/datasource`. Within these directories we should be aiming for pretty much 100% unit testing coverage, but the steps represent integration pipelines which are much less likely to be unit tested.

File: src/otg/preprocess/finngen/summary_stats.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s %(message)s")
+
+
+@dataclass
+class FinnGenSummaryStats(SummaryStatistics):
+    """Summary statistics dataset for FinnGen."""
+
+    @classmethod
+    def get_schema(cls: type[SummaryStatistics]) -> StructType:
```
Comment (d0choa) [inline src/otg/preprocess/finngen/summary_stats.py:30]: not needed. you get it by inheritance from SummaryStatistics

File: src/otg/preprocess/finngen/summary_stats.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        )
+
+        # Initializing summary statistics object:
+        return cls(
+            _df=processed_summary_stats_df,
+            _schema=cls.get_schema(),
+        )
+
+
+def ingest_finngen_summary_stats(
```
Comment (d0choa) [inline src/otg/preprocess/finngen/summary_stats.py:83]: see explanation on studyIndex. Better to move this to `otg/finngen.py`.

File: src/otg/preprocess/finngen/summary_stats.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+):
+    """Summary stats ingestion for FinnGen.
+
+    Args:
+        finngen_study_id (str): ID of the individual study being ingested.
+        finngen_summary_stats_location (str): Google Storage URI for the summary stats to ingest.
+        finngen_summary_stats_out (str): Output path for the FinnGen summary statistics dataset.
+        spark_write_mode (str): Dataframe write mode.
+    """
+    spark = SparkSession.builder.master("yarn").appName("ingest_finngen").getOrCreate()
```
Comment (d0choa) [inline src/otg/preprocess/finngen/summary_stats.py:97]: same comment as studyIndex

File: src/otg/preprocess/finngen/summary_stats.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        spark.read.option("delimiter", "\t")
+        .csv(finngen_summary_stats_location, header=True)
+        .repartition("#chrom")
+    )
+
+    # Process and output the data.
+    out_filename = f"{finngen_summary_stats_out}"
+    FinnGenSummaryStats.from_finngen_harmonized_summary_stats(
+        summary_stats_df, finngen_study_id
+    ).df.sortWithinPartitions("position").write.partitionBy("chromosome").mode(
```
Comment (d0choa) [inline src/otg/preprocess/finngen/summary_stats.py:113]: @DSuveges did you solve your issues with this?

File: Makefile [hunk 9]
```diff
@@ -68,4 +68,5 @@ build: clean ## Build Python package with dependencies
 	@gsutil cp ./dist/${APP_NAME}-${VERSION_NO}-py3-none-any.whl ${BUCKET_NAME}
 	@gsutil cp ./dist/config.tar.gz ${BUCKET_NAME}
 	@gsutil cp ./utils/install_dependencies_on_cluster.sh ${BUCKET_NAME}
+	@gsutil -m cp -r ./src/otg/preprocess ${BUCKET_NAME}
```
Comment (d0choa) [inline Makefile:71]: This change might not belong to this PR

File: src/otg/preprocess/finngen/summary_stats.py (hunk 8 above)
Comment (DSuveges) [reply src/otg/preprocess/finngen/summary_stats.py]: Oh, yes, I would remove `partitionBy("chromosome")` from after write.

File: src/otg/preprocess/finngen/study_index.py (hunk 2 above)
Comment (tskir) [reply src/otg/preprocess/finngen/study_index.py]: Thank you for spotting this, I've now reconciled all changes and added those as well.

File: src/otg/preprocess/finngen/study_index.py (hunk 3 above)
Comment (tskir) [reply src/otg/preprocess/finngen/study_index.py]: Thank you! This was an abstract method before, and now this is indeed not required. Fixed.

File: src/otg/preprocess/finngen/study_index.py (hunk 4 above)
Comment (tskir) [reply src/otg/preprocess/finngen/study_index.py]: Agreed. The layout of this PR is now in line with existing architecture, so fixed.

File: src/otg/preprocess/finngen/summary_stats.py (hunk 5 above)
Comment (tskir) [reply src/otg/preprocess/finngen/summary_stats.py]: You're right, fixed.

File: src/otg/preprocess/finngen/summary_stats.py (hunk 6 above)
Comment (tskir) [reply src/otg/preprocess/finngen/summary_stats.py]: Agreed, now fixed.

File: src/otg/preprocess/finngen/study_index.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+    Args:
+        finngen_phenotype_table_url (str): FinnGen API for fetching the list of studies.
+        finngen_release_prefix (str): Release prefix pattern.
+        finngen_summary_stats_url_prefix (str): URL prefix for summary statistics location.
+        finngen_summary_stats_url_suffix (str): URL prefix suffix for summary statistics location.
+        finngen_study_index_out (str): Output path for the FinnGen study index dataset.
+        spark_write_mode (str): Dataframe write mode.
+    """
+    spark = SparkSession.builder.master("yarn").appName("ingest_finngen").getOrCreate()
```
[the remark this answers is not in the document]
Comment (tskir) [reply src/otg/preprocess/finngen/study_index.py]: Fixed.

File: src/otg/preprocess/finngen/summary_stats.py (hunk 7 above)
Comment (tskir) [reply src/otg/preprocess/finngen/summary_stats.py]: Fixed.

File: Makefile (hunk 9 above)
Comment (tskir) [reply Makefile]: You're absolutely right, now fixed.

File: src/otg/preprocess/finngen/summary_stats.py (hunk 8 above)
Comment (tskir) [reply src/otg/preprocess/finngen/summary_stats.py]: This has been now fixed in the following way:
* The entire output is one big Parquet dataset, partitioned by study and then by chromosome.
* The individual *.parquet files will not contain either `study` or `chromosome` columns.
* However, when you ingest the entire dataset (& then filter by study and/or chromosome as required), the columns will be properly populated.

**Review by DSuveges — APPROVED** [review]
OK. All look good now.
